### PR TITLE
SonarQube build summary now includes quality gate status.

### PR DESCRIPTION
**Pull request by HSAR**

### CODE REVIEW

File: Tasks/Gradle/gradletask.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ import fs = require('fs');
 import path = require('path');
 
 // Lowercased file names are to lessen the likelihood of xplat issues
-import sqCommon = require('sonarqube-common/sonarqube-common');
+import sqCommon = require('./sonarqube-common/sonarqube-common');
```
Comment (bgavrilMS) [inline Tasks/Gradle/gradletask.ts]: Consider renaming this directory to "SonarQube" and moving it into CodeAnalysis. Also, there is no need for all the file names to be prefixed with sonarqube-

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts [hunk 2]
```diff
@@ -180,22 +182,38 @@ function isNullOrEmpty(str) {
...
-        return tl.loc('sqAnalysis_BuildSummaryNotAvailableInPrBuild');
+        return Q.when(tl.loc('sqAnalysis_BuildSummaryNotAvailableInPrBuild'));
     }
 
-    var taskReport: TaskReport = getSonarQubeTaskReport(sqBuildFolder);
-    if (!taskReport) {
-        throw new Error(tl.loc('sqAnalysis_TaskReportInvalid'));
-    }
+    // Wait for analysis to finish so that we can get analysis information from the SonarQube server
+    var timeout:number = 300;
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Unless you are going to expose these settings to the user, there's no reason to have them here. The SonarQubeMetrics object is more than capable of providing its own default values for these.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts [hunk 3]
```diff
@@ -180,22 +182,38 @@ function isNullOrEmpty(str) {
...
+    var delay:number = 1;
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Unless you are going to expose these settings to the user, there's no reason to have them here. The SonarQubeMetrics object is more than capable of providing its own default values for these.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts [hunk 4]
```diff
@@ -180,22 +182,38 @@ function isNullOrEmpty(str) {
...
 
-    var linkToDashBoard: string = createLinkToSonarQubeDashboard(sqBuildFolder);
+    var sqRunSettings:SonarQubeRunSettings = getSonarQubeTaskReport(sqBuildFolder);
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Logically, the task of finding and reading the report-task.txt file lays with the SonarQubeRunSettings object. You can either do it via the ctor or via a static CreateFromFile factory method. I'd go for a ctor. https://blogs.msdn.microsoft.com/kcwalina/2004/10/11/design-guidelines-update-factories-vs-constructors/

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts [hunk 5]
```diff
@@ -205,33 +223,37 @@ function getOrCreateSonarQubeStagingDirectory(): string {
...
-function createLinkToSonarQubeDashboard(sqBuildFolder: string): string {
-    var taskReport: TaskReport = getSonarQubeTaskReport(sqBuildFolder);
-    if (!taskReport) {
-        throw new Error(tl.loc('sqAnalysis_TaskReportInvalid'));
+/**
+ * Creates a string containing Markdown of the SonarQube quality gate status for this project.
+ * @param sqRunSettings SonarQubeRunSettings object for the applicable run
+ * @returns {string}    A Markdown-formatted report on the SonarQube quality gate status
+ */
+function createQualityGateStatusSection(sqReportBuilder: SonarQubeReportBuilder): Q.Promise<string> {
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: The intent is not clear. Have a different method that describes the hook we are using for test. Then chain the fetch method to it.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts [hunk 6]
```diff
@@ -180,22 +182,38 @@ function isNullOrEmpty(str) {
...
+    var sqRunSettings:SonarQubeRunSettings = getSonarQubeTaskReport(sqBuildFolder);
 
-    // Put the quality gate status and dashboard link sections together with the Markdown newline
-    return linkToDashBoard;
+    var sqServer:ISonarQubeServer = new SonarQubeServer();
+    var analysisMetrics:SonarQubeMetrics = new SonarQubeMetrics(sqServer, sqRunSettings.ceTaskId, timeout, delay);
+    var sqReportBuilder:SonarQubeReportBuilder = new SonarQubeReportBuilder(sqRunSettings, analysisMetrics);
+
+    // Looks like: "[Detailed SonarQube report >](https://mySQserver:9000/dashboard/index/foo "foo Dashboard")"
+    var linkToDashBoard:string = sqReportBuilder.createLinkToSonarQubeDashboard();
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: The sections of the report and the binding of these sections should also be part of the ReportBuilder and not exposed here.

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        return this.endpoint;
+    }
+
+    /**
+     * Invokes a REST endpoint on the SonarQube server.
+     * @param path The host-relative path to the REST endpoint (e.g. /api/ce/task)
+     * @returns A promise, resolving with a JSON object representation of the response. Rejects on error or non-200 header.
+     */
+    public callSonarQubeRestEndpoint(path:string):Q.Promise<Object> {
+        var defer = Q.defer<Object>();
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: We have way too much logic in that soanrqube-common.ts file. To break it up, create a separate file for the endpoint and move all the logic there. It's not related to your work, but http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    getEndpoint():SonarQubeEndpoint;
+}
+
+export class SonarQubeServer implements ISonarQubeServer {
+
+    private endpoint:SonarQubeEndpoint;
+
+    constructor() {
+        this.endpoint = sqCommon.getSonarQubeEndpoint();
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: We have way too much logic in that soanrqube-common.ts file. To break it up, create a separate file for the endpoint and move all the logic there. It's not related to your work, but http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    public getEndpoint():SonarQubeEndpoint {
+        return this.endpoint;
+    }
+
+    /**
+     * Invokes a REST endpoint on the SonarQube server.
+     * @param path The host-relative path to the REST endpoint (e.g. /api/ce/task)
+     * @returns A promise, resolving with a JSON object representation of the response. Rejects on error or non-200 header.
+     */
+    public callSonarQubeRestEndpoint(path:string):Q.Promise<Object> {
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: How about "invokeGetApi"

**Comment by HSAR** [comment]
Todo: Remove nock-related stuff which won't be needed since we're now mocking ISonarQubeServer.

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+export class SonarQubeServer implements ISonarQubeServer {
+
+    private endpoint:SonarQubeEndpoint;
+
+    constructor() {
+        this.endpoint = sqCommon.getSonarQubeEndpoint();
+    }
+
+    public getEndpoint():SonarQubeEndpoint {
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Does this need to be public?

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts (hunk 8 above)
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Also, the server should be given the endpoint, it shouldn't have to know how to get it.

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+export interface ISonarQubeServer {
+    callSonarQubeRestEndpoint(path:string):Q.Promise<Object>;
+
+    getEndpoint():SonarQubeEndpoint;
+}
+
+export class SonarQubeServer implements ISonarQubeServer {
+
+    private endpoint:SonarQubeEndpoint;
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Could you move all the endpoint logic to a separate file? We really need to clean up that sonarqube-common.ts file http://programmer.97things.oreilly.com/wiki/index.php/The_Boy_Scout_Rule

File: Tasks/Maven/sonarqube-common/sonarqube-metrics.ts [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+     * @param server  Server the task was sent to
+     * @param taskId  A string identifying the task
+     * @param timeout (optional) Time, in seconds, to wait for the analysis job to finish before returning the promise as false.
+     * @param delay   (optional) Time, in seconds, to wait between polling attempts. Warning: Does not also wait for previous request to finish.
+     */
+    constructor(server:ISonarQubeServer, taskId:string, timeout?:number, delay?:number) {
+        this.server = server;
+        this.taskId = taskId;
+
+        this.timeout = timeout|| 60;
```
Comment (bgavrilMS) [inline Tasks/Maven/sonarqube-common/sonarqube-metrics.ts]: 300

File: Tasks/Gradle/sonarqube-common/sonarqube-reportbuilder.ts [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        this.taskMetrics = taskMetrics;
+    }
+
+
+    // Creates a string containing Markdown of a link to the SonarQube dashboard for this project.
+    /**
+     * Creates a Markdown-formatted link to the SonarQube dashboard for this project.
+     * @returns {string} A single-line, Markdown-formatted link to the SonarQube dashboard for this project
+     */
+    public createLinkToSonarQubeDashboard(): string {
```
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-reportbuilder.ts]: As stated above, the report builder should only expose a single public method. There is no need to expose the whole "sections" logic

**Comment by bgavrilMS** [comment]
LGTM

File: Tasks/Gradle/gradletask.ts (hunk 1 above)
Comment (HSAR) [inline Tasks/Gradle/gradletask.ts]: Accepted, in progress.

Comment (HSAR) [inline Tasks/Gradle/gradletask.ts]: Resolved, done.

I considered leaving them as "sonarqube-server" etc. to match the class names, but that makes a full-name import look like "sonarqube-server.SonarQubeServer" which is just redundant.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts (hunk 3 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Accepted, in progress.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Resolved, folded them into default parameters in SonarQubeMetrics. We may open them up to user configuration later, but for now best to keep it closed.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts (hunk 4 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: I don't agree. Referencing the design of PMD on Maven - mavenpmd handles the dirty work of finding and locating the pmd.xml file, from which AnalysisResult is created.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts (hunk 6 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Accepted - this makes much more sense.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Resolved, done.

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts (hunk 10 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Accepted - I initially thought that another class might need it, but YAGNI applies.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Resolved, done.

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts (hunk 8 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Accepted, in progress.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Resolved, done.

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts (hunk 11 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Accepted, in progress.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Resolved, done.

File: Tasks/Maven/sonarqube-common/sonarqube-metrics.ts (hunk 12 above)
Comment (HSAR) [inline Tasks/Maven/sonarqube-common/sonarqube-metrics.ts]: Accepted.

Comment (HSAR) [inline Tasks/Maven/sonarqube-common/sonarqube-metrics.ts]: Resolved - changed in the move to default parameters.

File: Tasks/Gradle/sonarqube-common/sonarqube-reportbuilder.ts (hunk 13 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-reportbuilder.ts]: Accepted.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-reportbuilder.ts]: Resolved as part of fixing the above point.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts (hunk 5 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: I don't understand the point you were making here - let's discuss this offline.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts (hunk 4 above)
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: report-task.txt is build agnostic. Its location is build specifc. Even so, we should have little or no code in that common.ts file. We'll refactor as we go along. This seems like an easy win.

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts (hunk 9 above)
Comment (bgavrilMS) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: As you wish, invokeRestApi is the name I used in PS.

File: Tests/L0/Maven/_suite.ts [hunk 14]
```diff
@@ -1552,4 +1542,111 @@ describe('Maven Suite', function () {
...
+        // Arrange
+        var mockRunSettings:SonarQubeRunSettings = new SonarQubeRunSettings("projectKey", "serverUrl", "http://dashboardUrl", "asdfghjklqwertyuiopz", "taskUrl");
+        var mockServer:MockSonarQubeServer = new MockSonarQubeServer();
+
+        var analysisMetrics:SonarQubeMetrics = new SonarQubeMetrics(mockServer, mockRunSettings.ceTaskId, 10, 1); // override to a 10-second timeout
+        var sqReportBuilder:SonarQubeReportBuilder = new SonarQubeReportBuilder(mockRunSettings, analysisMetrics);
+
+        // Mock responses from the server for the task and analysis details
+        var taskDetailsJsonObject:any = JSON.parse(fs.readFileSync(path.join(__dirname, 'data/task_details.json'), 'utf-8'));
+        mockServer.setupMockApiCall('/api/ce/task?id=asdfghjklqwertyuiopz', taskDetailsJsonObject);
```
Comment (bgavrilMS) [inline Tests/L0/Maven/_suite.ts]: setupMockApiCall could take a file path instead of the deserialized JSON. It will make the test easier to read.

**Comment by bgavrilMS** [comment]
LGTM! The test are very easy to understand, much simpler than the L0 tests

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts (hunk 4 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Accepted. Refactoring into SQRunSettings as a factory method.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Resolved, done.

File: Tests/L0/Maven/_suite.ts (hunk 14 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: The reason I kept the object can be seen in one of the tests where I deliberately return an empty object as a check for tolerance of invalid input.

File: Tasks/Gradle/sonarqube-common/sonarqube-common.ts (hunk 5 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Discussed offline - sqAnalysisWaitForAnalysis is actually user input.

Accepted: change to unify UI with MSBuild SonarQube integration.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-common.ts]: Resolved, done.

File: Tasks/Gradle/sonarqube-common/sonarqube-server.ts (hunk 9 above)
Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Accepted, in progress.

Comment (HSAR) [inline Tasks/Gradle/sonarqube-common/sonarqube-server.ts]: Resolved, done.

**Pull request merged**